File: 6.projector/ts/src/projector.ts

import { Config } from "./config"
import * as fs from "fs";
import * as path from "path";

export type Data = {
    projector: {
        [key: string]: { // pwd
            [key: string]: string // key -> value
        }
    }
}

const defaultData = {
    projector: {}
}

export default class Projector {

    constructor(private config: Config, private data: Data) { }

    save() {
        const configPath = path.dirname(this.config.config);
        if (!fs.existsSync(configPath)) {
            fs.mkdirSync(configPath, { recursive: true });
        }
        // just overwrite the file if it's exists
        fs.writeFileSync(this.config.config, JSON.stringify(this.data));
    }

    getValueAll(): { [key: string]: string } {
        let curr = this.config.pwd;
        let prev = "";

        const paths: string[] = [];
        do {
            prev = curr;
            paths.push(curr);
            curr = path.dirname(curr);
        } while (curr != prev);

        return paths.reverse().reduce((acc, path) => {
            const value = this.data.projector[path];
            if (value) { Object.assign(acc, value) }

            return acc;
        }, {})
    }
    getValue(key: string): string | undefined {
        let curr = this.config.pwd;
        let prev = "";

        let out: string | undefined = undefined;
        do {

            const value = this.data.projector[curr]?.[key];
            if (value) { out = value; break; }

            prev = curr;
            curr = path.dirname(curr);

        } while (curr != prev);

        return out;
    }
    setValue(key: string, value: string) {
        let dir = this.data.projector[this.config.pwd];
        if (!dir) {
            dir = this.data.projector[this.config.pwd] = {};
        }
        dir[key] = value;
    }
    removeValue(key: string) {
        const dir = this.data.projector[this.config.pwd];
        if (dir) {
            delete dir[key];
        }
    }

    static fromConfig(config: Config): Projector {
        if (fs.existsSync(config.config)) {

            let data: Data /* | undefined */ /* = undefined */;
            try {
                data = JSON.parse(fs.readFileSync(config.config).toString());
            } catch (e) {
                data = defaultData;
            }
            return new Projector(config, data);
        }
        return new Projector(config, defaultData);
    }
}

File: 6.projector/rs/src/bin/projector.rs
use anyhow::Result;
use clap::Parser;
use rs::{
    config::{self, Operation},
    opts::Opts,
    projector::Projector,
};

fn main() -> Result<()> {
    let opts = Opts::parse();
    // println!("{:?}", opts)

    // let operation: Operation = opts.args.try_into()?;
    // let config = config::get_config(opts.config)?;
    // let pwd = config::get_pwd(opts.pwd)?;
    // println!("op: {:?}; config: {:?}; pwd: {:?}", operation, config, pwd);
    let conf = config::Config::try_from(opts)?;
    // println!("{:?}", config);

    let mut proj = Projector::from_config(conf.config, conf.pwd);

    match conf.operation {
        Operation::Print(None) => {
            let value = proj.get_value_all();
            let value = serde_json::to_string(&value)?;
            println!("{}", value);
        }
        Operation::Print(Some(k)) => {
            let value = proj.get_value(&k);
            // if let Some(value) = value {
            //     println!("{}", value);
            // }
            value.iter().for_each(|v| println!("{}", v));
        }
        Operation::Add(k, v) => {
            proj.set_value(k.to_string(), v.to_string());
            proj.save()?;
        }
        Operation::Remove(k) => {
            proj.remove_value(&k);
            proj.save()?;
        }
    }
    Ok(())
}
